### PR TITLE
refactor WIDERFaceDataset

### PR DESCRIPTION
**Pull request by MambaWong**
refactor WIDERFaceDataset

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Hi @MambaWong ,
Thanks for your kind PR. We have an ongoing PR in https://github.com/open-mmlab/mmdetection/pull/8468, which needs to be rebased later. We prefer to merge your PR. Therefore, would you like to migrate the changes and comments from #8468 so that we could merge your PR?

File: mmdet/datasets/wider_face.py [hunk 1]
```diff
@@ -15,40 +16,44 @@ class WIDERFaceDataset(XMLDataset):
     Conversion scripts can be found in
     https://github.com/sovrasov/wider-face-pascal-voc-annotations
     """
-    CLASSES = ('face', )
+    METAINFO = {'CLASSES': ('face', ), 'PALETTE': [(0, 255, 0)]}
 
-    PALETTE = [(0, 255, 0)]
+    def __init__(self, img_subdir: str = '', **kwargs):
+        super(WIDERFaceDataset, self).__init__(img_subdir, **kwargs)
```
Comment (BIGWangYuDong) [inline mmdet/datasets/wider_face.py:22]:
```suggestion
        super().__init__(img_subdir=img_subdir, **kwargs)
```

File: mmdet/datasets/wider_face.py [hunk 2]
```diff
@@ -15,40 +18,57 @@ class WIDERFaceDataset(XMLDataset):
...
     https://github.com/sovrasov/wider-face-pascal-voc-annotations
     """
-    CLASSES = ('face', )
-
-    PALETTE = [(0, 255, 0)]
+    METAINFO = {'CLASSES': ('face', ), 'PALETTE': [(0, 255, 0)]}
 
     def __init__(self, **kwargs):
-        super(WIDERFaceDataset, self).__init__(**kwargs)
+        super(WIDERFaceDataset, self).__init__(img_subdir='', **kwargs)
```
Comment (BIGWangYuDong) [inline mmdet/datasets/wider_face.py:24]:
```suggestion
        super().__init__(img_subdir='', **kwargs)
```

File: mmdet/datasets/xml_style.py [hunk 3]
```diff
@@ -99,14 +94,32 @@ def parse_data_info(self, img_info: dict) -> Union[dict, List[dict]]:
...
         data_info['width'] = width
 
+        data_info['instances'] = self._parse_instance_info(
+            raw_ann_info, minus_one=True)
+
+        return data_info
+
+    def _parse_instance_info(self,
+                             raw_ann_info: ET,
+                             minus_one: bool = False) -> List[dict]:
```
Comment (BIGWangYuDong) [inline mmdet/datasets/xml_style.py:112]: where will use minus_one = False?

File: mmdet/datasets/xml_style.py [hunk 4]
```diff
@@ -99,14 +94,32 @@ def parse_data_info(self, img_info: dict) -> Union[dict, List[dict]]:
...
+            img_bytes = self.file_client.get(data_info['img_path'])
             img = mmcv.imfrombytes(img_bytes, backend='cv2')
             width, height = img.shape[:2]
             del img, img_bytes
 
         data_info['height'] = height
         data_info['width'] = width
 
+        data_info['instances'] = self._parse_instance_info(
+            raw_ann_info, minus_one=True)
```
Comment (BIGWangYuDong) [inline mmdet/datasets/xml_style.py:106]: minus_one is a hardcode here

File: mmdet/datasets/voc.py [hunk 5]
```diff
@@ -22,7 +22,7 @@ class VOCDataset(XMLDataset):
     }
 
     def __init__(self, **kwargs):
-        super().__init__(**kwargs)
+        super().__init__(minus_one=True, **kwargs)
```
Comment (BIGWangYuDong) [inline mmdet/datasets/voc.py:25]: Seems this is a hard code. How about set a default value?

File: mmdet/datasets/wider_face.py [hunk 6]
```diff
@@ -15,40 +18,62 @@ class WIDERFaceDataset(XMLDataset):
...
     """
-    CLASSES = ('face', )
-
-    PALETTE = [(0, 255, 0)]
+    METAINFO = {'CLASSES': ('face', ), 'PALETTE': [(0, 255, 0)]}
 
     def __init__(self, **kwargs):
-        super(WIDERFaceDataset, self).__init__(**kwargs)
+        super(WIDERFaceDataset, self).__init__(
+            img_subdir='', minus_one=False, **kwargs)
```
Comment (BIGWangYuDong) [inline mmdet/datasets/wider_face.py:25]: Same hard code

File: mmdet/datasets/voc.py (hunk 5 above)
Comment (MambaWong) [reply mmdet/datasets/voc.py:25]: @chhluo Can you take a look at it? And give some suggestions.

File: configs/_base_/datasets/wider_face.py [hunk 7]
```diff
@@ -1,63 +1,60 @@
...
-        ann_file=data_root + 'val.txt',
-        img_prefix=data_root + 'WIDER_val/',
+        data_root=data_root,
+        ann_file='val.txt',
+        data_prefix=dict(img='WIDER_val'),
+        test_mode=True,
         pipeline=test_pipeline))
+test_dataloader = val_dataloader
+
+val_evaluator = dict(type='VOCMetric', metric='mAP', eval_mode='11points')
```
Comment (BIGWangYuDong) [inline configs/_base_/datasets/wider_face.py:59]: 11points or area?

Comment (ZwwWayne) [reply configs/_base_/datasets/wider_face.py:59]: cannot get the meaning of this coments

**Pull request closed without merging**